File: app/src/main/java/cz/vitskalicky/lepsirozvrh/UpdateBroadcastReciever.java
package cz.vitskalicky.lepsirozvrh;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import cz.vitskalicky.lepsirozvrh.bakaAPI.rozvrh.RozvrhAPI;
import cz.vitskalicky.lepsirozvrh.notification.PermanentNotification;
import cz.vitskalicky.lepsirozvrh.widget.WidgetProvider;

/**
 * Broadcast receiver that updates notification and widgets when receives a broadcast.
 */
public class UpdateBroadcastReciever extends BroadcastReceiver {
    private static final String TAG = UpdateBroadcastReciever.class.getSimpleName();
    public static final int REQUEST_CODE = 64857;
    /**
     * +1 for next, -1 for prev
     */
    public static final String EXTRA_NEXT_PREV = BuildConfig.APPLICATION_ID + ".extra-next-or-prev-lesson";
    public static final String ACTION_NEXT_PREV = BuildConfig.APPLICATION_ID + ".action-next-or-prev-lesson";

    Context context;
    RozvrhAPI rozvrhAPI = null;
    MainApplication application;
    PendingResult pendingResult;

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "Broadcast recieved");
        this.context = context;
        rozvrhAPI = AppSingleton.getInstance(context).getRozvrhAPI();
        application = (MainApplication) context.getApplicationContext();
        pendingResult = goAsync();

        if (intent != null && intent.getAction() != null && intent.getAction().equals(ACTION_NEXT_PREV) && intent.hasExtra(EXTRA_NEXT_PREV)){
            int offset = intent.getIntExtra(EXTRA_NEXT_PREV, 0);
            application.getNotificationState().setOffset(application.getNotificationState().getOffset() + offset);
            application.scheduleUpdate(application.getNotificationState().getOffsetResetTime());
        }

        rozvrhAPI.getRozvrh(Utils.getCurrentMonday(), rozvrhWrapper -> {
            PermanentNotification.update(rozvrhWrapper.getRozvrh(), application);
            WidgetProvider.updateAll(rozvrhWrapper.getRozvrh(), context);

            application.updateUpdateTime(pendingResult::finish);
        });
    }
}

File: app/src/main/java/cz/vitskalicky/lepsirozvrh/UpdateBroadcastReciever.kt
package cz.vitskalicky.lepsirozvrh

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import cz.vitskalicky.lepsirozvrh.model.relations.RozvrhRelated
import cz.vitskalicky.lepsirozvrh.notification.PermanentNotification.update
import cz.vitskalicky.lepsirozvrh.widget.WidgetProvider
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlin.coroutines.Continuation

/**
 * Broadcast receiver that updates notification and widgets when receives a broadcast.
 */
class UpdateBroadcastReciever : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        Log.d(TAG, "Broadcast received")
        val application = context.applicationContext as MainApplication
        val pendingResult = goAsync()
        if (intent.action != null && intent.action == ACTION_NEXT_PREV && intent.hasExtra(EXTRA_NEXT_PREV)) {
            val offset = intent.getIntExtra(EXTRA_NEXT_PREV, 0)
            application.notificationState.offset += offset
            application.scheduleUpdate(application.notificationState.offsetResetTime)
        }
        GlobalScope.launch {
            val rozvrh: RozvrhRelated? = application.repository.getRozvrh(Utils.getCurrentMonday(), false)
            update(rozvrh,application)
            WidgetProvider.updateAll(rozvrh, context)
            application.updateUpdateTime()
            pendingResult.finish()
        }
    }

    companion object {
        private val TAG = UpdateBroadcastReciever::class.java.simpleName
        const val REQUEST_CODE = 64857

        /**
         * +1 for next, -1 for prev
         */
        const val EXTRA_NEXT_PREV = BuildConfig.APPLICATION_ID + ".extra-next-or-prev-lesson"
        const val ACTION_NEXT_PREV = BuildConfig.APPLICATION_ID + ".action-next-or-prev-lesson"
    }
}